graph/telemetry/istio/appender/health.go

```diff
@@ -76,160 +87,166 @@ func (a *HealthAppender) attachHealth(trafficMap graph.TrafficMap, globalInfo *g
 			continue
 		}
 
+		var req healthRequest
+		var ok bool
+		if req, ok = healthReqs[n.Namespace]; !ok {
+			req = healthRequest{}
+		}
+
 		switch n.NodeType {
 		case graph.NodeTypeApp:
-			if _, nsOk := healthReqs[n.Namespace]; !nsOk {
-				healthReqs[n.Namespace] = make(map[string][]*graph.Node)
-			}
 			// always get app health for app node (used for app box health)
-			healthReqs[n.Namespace][graph.NodeTypeApp] = append(healthReqs[n.Namespace][graph.NodeTypeApp], n)
+			req.app = true
 
 			// for versioned app node, get workload health as well (used for the versioned app node itself)
 			if graph.IsOK(n.Workload) {
-				healthReqs[n.Namespace][graph.NodeTypeWorkload] = append(healthReqs[n.Namespace][graph.NodeTypeWorkload], n)
+				req.workload = true
 			}
 		case graph.NodeTypeWorkload:
-			if _, nsOk := healthReqs[n.Namespace]; !nsOk {
-				healthReqs[n.Namespace] = make(map[string][]*graph.Node)
-			}
-
-			healthReqs[n.Namespace][graph.NodeTypeWorkload] = append(healthReqs[n.Namespace][graph.NodeTypeWorkload], n)
+			req.workload = true
 		case graph.NodeTypeService:
-			if _, nsOk := healthReqs[n.Namespace]; !nsOk {
-				healthReqs[n.Namespace] = make(map[string][]*graph.Node)
-			}
-
-			healthReqs[n.Namespace][graph.NodeTypeService] = append(healthReqs[n.Namespace][graph.NodeTypeService], n)
+			req.service = true
 		}
+		
+		healthReqs[n.Namespace] = req
+		nodesWithHealth = append(nodesWithHealth, n)
 	}
 
-	// Execute health fetches and attach retrieved health data to nodes
 	bs := globalInfo.Business
 	ctx := globalInfo.Context
 
-	// Gather all the health data we'll need ahead of time key'd by namespace.
-	appHealth := make(map[string]models.NamespaceAppHealth)
-	serviceHealth := make(map[string]models.NamespaceServiceHealth)
-	workloadHealth := make(map[string]models.NamespaceWorkloadHealth)
+	var cancel context.CancelFunc
+	if ctx == nil {
+		ctx = context.Background()
+	}
+	// All requests shouldn't take longer than the requested duration.
```

Comment:
Was a little unsure about this. We need to choose some time period for the timeout and I figured if you are going past the requested duration then you are going to get stale results anyways since the frontend would likely hit the next refresh interval and fire off a new graph request.

Reply:
my only concern is that this could lead to never getting results. but maybe then a user could select a longer refresh.

Reply:
Yeah that could happen and there's probably not much to tell the user exactly what's going on and how they can fix it. If the frontend were to tell the user what the error was and how to fix it I think it'd be fine to have the `timeout = refresh interval`. For now, I can make the timeout 15 mins which is the largest refresh interval? Seems high but better than nothing.